weave/trace_server/reference/generate.py

```diff
@@ -0,0 +1,374 @@
+from __future__ import annotations
+
+from typing import Annotated, Callable, NamedTuple
+
+from fastapi import APIRouter, Depends, Form, Header, UploadFile
+from fastapi.responses import StreamingResponse
+
+from weave.trace_server import trace_server_interface as tsi
+
+SERVICE_TAG_NAME = "Service"
+CALLS_TAG_NAME = "Calls"
+OPS_TAG_NAME = "Ops"
+OBJECTS_TAG_NAME = "Objects"
+TABLES_TAG_NAME = "Tables"
+REFS_TAG_NAME = "Refs"
+FILES_TAG_NAME = "Files"
+FEEDBACK_TAG_NAME = "Feedback"
+COST_TAG_NAME = "Costs"
+COMPLETIONS_TAG_NAME = "Completions"
+ACTIONS_TAG_NAME = "Actions"
+
+
+class AuthParams(NamedTuple):
+    headers: dict[str, str] | None = None
+    cookies: dict[str, str] | None = None
+    auth: tuple[str, str] | None = None
+
+    def __hash__(self) -> int:
+        return hash(
+            (
+                tuple(self.headers.items()) if self.headers else None,
+                tuple(self.cookies.items()) if self.cookies else None,
+                self.auth,
+            )
+        )
+
+
+class NoopTraceServer(tsi.TraceServerInterface): ...
+
+
+def noop_trace_server_factory(auth: AuthParams) -> tsi.TraceServerInterface:
+    # This type-ignore is safe, it's just used to instantiate a stub implementation
+    # without having to redefine all of the methods (which would be pointless because
+    # this is a stub that does nothing).
+    return NoopTraceServer()  # type: ignore
+
+
+class ServerDependency:
+    """Factory for creating server dependencies with proper authorization."""
+
+    def __init__(
+        self,
+        server_factory: Callable[[AuthParams], tsi.TraceServerInterface] = (
+            noop_trace_server_factory
+        ),
+        auth_dependency: Callable[[], AuthParams] = lambda: AuthParams(),
+    ):
+        """
+        Initialize with auth dependencies and server factory.
+
+        Args:
+            endpoint_auth_mapping: Dict mapping endpoint names directly to auth dependencies
+            server_factory: Function that creates a server from auth params and endpoint name
+        """
+        self.auth_dependency = auth_dependency
+        self.server_factory = server_factory
+
+    def get_server(self) -> Callable[[AuthParams], tsi.TraceServerInterface]:
+        """Get a server dependency with the appropriate auth for the operation."""
+
+        def _get_server(
+            auth_params: AuthParams = Depends(self.auth_dependency),
+        ) -> tsi.TraceServerInterface:
+            return self.server_factory(auth_params)
+
+        return _get_server
+
+
+def generate_routes(
```

Comment:
The routes are basically all delegating to the server, which is dependency injected per request.

Defining all the routes here ensures that the relevant parts of the OpenAPI schema we generate for any implementation will be the same.  Parts that may differ include properties of the app itself, security, etc.